// solid-chartjs/dist/index.cjs
'use strict';

var web = require('solid-js/web');
var solidJs = require('solid-js');
var chart_js = require('chart.js');
var store = require('solid-js/store');

// src/chart.tsx
function chain(callbacks) {
  return (...args) => {
    for (const callback of callbacks)
      callback && callback(...args);
  };
}
function mergeRefs(...refs) {
  return chain(refs);
}
var _tmpl$ = /* @__PURE__ */ web.template(`<canvas>`);
function DefaultChart(props) {
  const [canvasRef, setCanvasRef] = solidJs.createSignal();
  const [chart, setChart] = solidJs.createSignal();
  const merged = solidJs.mergeProps({
    width: 512,
    height: 512,
    type: "line",
    data: {},
    options: {
      responsive: true
    },
    plugins: []
  }, props);
  const init = () => {
    const ctx = canvasRef()?.getContext("2d");
    const config = store.unwrap(merged);
    if (config.type !== "radar") {
      if (config.options.scales?.r) {
        delete config.options.scales?.r;
      }
    }
    const chart2 = new chart_js.Chart(ctx, {
      type: config.type,
      data: config.data,
      options: config.options,
      plugins: config.plugins
    });
    setChart(chart2);
  };
  solidJs.onMount(() => {
    init();
  });
  solidJs.createEffect(solidJs.on(() => merged.data, () => {
    chart().data = merged.data;
    chart().update();
  }, {
    defer: true
  }));
  solidJs.createEffect(solidJs.on(() => merged.options, () => {
    chart().options = merged.options;
    chart().update();
  }, {
    defer: true
  }));
  solidJs.createEffect(solidJs.on([() => merged.width, () => merged.height], () => {
    chart().resize(merged.width, merged.height);
  }, {
    defer: true
  }));
  solidJs.createEffect(solidJs.on(() => merged.type, () => {
    const dimensions = [chart().width, chart().height];
    chart().destroy();
    init();
    chart().resize(...dimensions);
  }, {
    defer: true
  }));
  solidJs.onCleanup(() => {
    chart()?.destroy();
    mergeRefs(props.ref, null);
  });
  return (() => {
    var _el$ = _tmpl$();
    var _ref$ = mergeRefs(props.ref, (el) => setCanvasRef(el));
    typeof _ref$ === "function" && web.use(_ref$, _el$);
    web.insert(_el$, () => merged.fallback);
    web.effect((_p$) => {
      var _v$ = merged.height, _v$2 = merged.width;
      _v$ !== _p$.e && web.setAttribute(_el$, "height", _p$.e = _v$);
      _v$2 !== _p$.t && web.setAttribute(_el$, "width", _p$.t = _v$2);
      return _p$;
    }, {
      e: void 0,
      t: void 0
    });
    return _el$;
  })();
}
function createTypedChart(type, registerables) {
  chart_js.Chart.register(registerables);
  return (props) => web.createComponent(DefaultChart, web.mergeProps({
    type
  }, props));
}
exports.Line = /* @__PURE__ */ createTypedChart("line", [chart_js.LineController, chart_js.CategoryScale, chart_js.LinearScale, chart_js.PointElement, chart_js.LineElement]);
exports.Bar = /* @__PURE__ */ createTypedChart("bar", [chart_js.BarController, chart_js.CategoryScale, chart_js.BarElement]);
exports.Radar = /* @__PURE__ */ createTypedChart("radar", [chart_js.RadarController, chart_js.RadialLinearScale]);
exports.Doughnut = /* @__PURE__ */ createTypedChart("doughnut", [chart_js.DoughnutController, chart_js.ArcElement]);
exports.PolarArea = /* @__PURE__ */ createTypedChart("polarArea", chart_js.PolarAreaController);
exports.Bubble = /* @__PURE__ */ createTypedChart("bubble", chart_js.BubbleController);
exports.Pie = /* @__PURE__ */ createTypedChart("pie", chart_js.PieController);
exports.Scatter = /* @__PURE__ */ createTypedChart("scatter", chart_js.ScatterController);

exports.DefaultChart = DefaultChart;
