// solid-chartjs/dist/index.js
import { use, insert, effect, setAttribute, template, createComponent, mergeProps as mergeProps$1 } from 'solid-js/web';
import { createSignal, mergeProps, onMount, createEffect, on, onCleanup } from 'solid-js';
import { Chart, LineController, CategoryScale, LinearScale, PointElement, LineElement, BarController, BarElement, RadarController, RadialLinearScale, DoughnutController, ArcElement, PolarAreaController, BubbleController, PieController, ScatterController } from 'chart.js';
import { unwrap } from 'solid-js/store';

// src/chart.tsx
function chain(callbacks) {
  return (...args) => {
    for (const callback of callbacks)
      callback && callback(...args);
  };
}
function mergeRefs(...refs) {
  return chain(refs);
}
var _tmpl$ = /* @__PURE__ */ template(`<canvas>`);
function DefaultChart(props) {
  const [canvasRef, setCanvasRef] = createSignal();
  const [chart, setChart] = createSignal();
  const merged = mergeProps({
    width: 512,
    height: 512,
    type: "line",
    data: {},
    options: {
      responsive: true
    },
    plugins: []
  }, props);
  const init = () => {
    const ctx = canvasRef()?.getContext("2d");
    const config = unwrap(merged);
    if (config.type !== "radar") {
      if (config.options.scales?.r) {
        delete config.options.scales?.r;
      }
    }
    const chart2 = new Chart(ctx, {
      type: config.type,
      data: config.data,
      options: config.options,
      plugins: config.plugins
    });
    setChart(chart2);
  };
  onMount(() => {
    init();
  });
  createEffect(on(() => merged.data, () => {
    chart().data = merged.data;
    chart().update();
  }, {
    defer: true
  }));
  createEffect(on(() => merged.options, () => {
    chart().options = merged.options;
    chart().update();
  }, {
    defer: true
  }));
  createEffect(on([() => merged.width, () => merged.height], () => {
    chart().resize(merged.width, merged.height);
  }, {
    defer: true
  }));
  createEffect(on(() => merged.type, () => {
    const dimensions = [chart().width, chart().height];
    chart().destroy();
    init();
    chart().resize(...dimensions);
  }, {
    defer: true
  }));
  onCleanup(() => {
    chart()?.destroy();
    mergeRefs(props.ref, null);
  });
  return (() => {
    var _el$ = _tmpl$();
    var _ref$ = mergeRefs(props.ref, (el) => setCanvasRef(el));
    typeof _ref$ === "function" && use(_ref$, _el$);
    insert(_el$, () => merged.fallback);
    effect((_p$) => {
      var _v$ = merged.height, _v$2 = merged.width;
      _v$ !== _p$.e && setAttribute(_el$, "height", _p$.e = _v$);
      _v$2 !== _p$.t && setAttribute(_el$, "width", _p$.t = _v$2);
      return _p$;
    }, {
      e: void 0,
      t: void 0
    });
    return _el$;
  })();
}
function createTypedChart(type, registerables) {
  Chart.register(registerables);
  return (props) => createComponent(DefaultChart, mergeProps$1({
    type
  }, props));
}
var Line = /* @__PURE__ */ createTypedChart("line", [LineController, CategoryScale, LinearScale, PointElement, LineElement]);
var Bar = /* @__PURE__ */ createTypedChart("bar", [BarController, CategoryScale, BarElement]);
var Radar = /* @__PURE__ */ createTypedChart("radar", [RadarController, RadialLinearScale]);
var Doughnut = /* @__PURE__ */ createTypedChart("doughnut", [DoughnutController, ArcElement]);
var PolarArea = /* @__PURE__ */ createTypedChart("polarArea", PolarAreaController);
var Bubble = /* @__PURE__ */ createTypedChart("bubble", BubbleController);
var Pie = /* @__PURE__ */ createTypedChart("pie", PieController);
var Scatter = /* @__PURE__ */ createTypedChart("scatter", ScatterController);

export { Bar, Bubble, DefaultChart, Doughnut, Line, Pie, PolarArea, Radar, Scatter };
